crates/provider-keyvalue-redis/src/lib.rs

```diff
@@ -74,12 +84,34 @@ impl KvRedisProvider {
             .await
             .context("failed to run provider")?;
         let connection = get_connection();
-        serve(
+        let invocations = bindings::serve(
             &connection.get_wrpc_client(connection.provider_key()),
             provider,
-            shutdown,
         )
         .await
+        .context("failed to serve exports")?;
+        let mut invocations = stream::select_all(invocations.into_iter().map(
```

Comment:
yeah, I thought about it copy-pasting these lines over and over again, I'll see what I can do

Reply:
See https://github.com/wasmCloud/wasmCloud/pull/2459/commits/1a492f4a17c2798c5274f4e0a18da605ad1c9a75

unfortunately, we can't quite smoosh the run and serve functions together, because Rust compiler is not smart enough yet to figure out the lifetimes required on the returned future

Reply:
That commit is still much cleaner so I think that is great for now!